README.md
Combo Handler
=============

[![Build Status](https://travis-ci.org/rgrove/combohandler.png?branch=master)](https://travis-ci.org/rgrove/combohandler)

This is a simple combo handler for Node.js, usable either as [Connect][]
middleware or as an [Express][] server. It works just like the combo handler
service on the Yahoo! CDN, which you'll be familiar with if you've used YUI.

The combo handler is compatible with the [YUI][] Loader, so you can use it to
host YUI, or you can use it with any other JavaScript or CSS if you're willing
to construct the combo URLs yourself.

The combo handler itself doesn't perform any caching or compression, but stick
[Nginx][] or something in front of it and you should be ready to rock in
production.

[Connect]: https://github.com/senchalabs/connect
[Express]: https://github.com/visionmedia/express
[Nginx]: http://nginx.org/
[YUI]: http://yuilibrary.com/


Installation
------------

Install using npm:

```bash
npm install combohandler
```

Or just clone the [GitHub repo](https://github.com/rgrove/combohandler):

```bash
git clone git://github.com/rgrove/combohandler.git
```


Usage
-----

The `combohandler` module provides a configurable Connect middleware that can be
used to add combo handling capability to any Connect-based request handler (like
Express).

The `combohandler/lib/server` module creates a standalone Express server instance,
or augments an existing server, to perform combo handling for a set of
configurable routes.


### As Express middleware

The combo handler middleware can be used as application-wide middleware for all
routes:

```js
var combo = require('combohandler');
app.use(combo.combine({rootPath: '/local/path/to/files'}));
```

Or as route middleware for a specific route:

```js
app.get('/foo', combo.combine({rootPath: '/local/path/to/foo'}), combo.respond);
```

In either case, the middleware will perform combo handling for files under the
specified local `rootPath` when requested using a URL with one or more file paths
in the query string:

```text
http://example.com/<route>?<path>[&path][...]
```

For example:

```text
http://example.com/foo?file1.js
http://example.com/foo?file1.js&file2.js
http://example.com/foo?file1.js&file2.js&subdir/file3.js
```

Attempts to traverse above the `rootPath` or to request a file that doesn't
exist will result in a `BadRequest` error being bubbled up.

Here's a basic Express app that uses the combo handler as route middleware for
multiple routes with different root paths:

```js
var combo   = require('combohandler'),
    express = require('express'),

    app = express();

app.configure(function () {
  app.use(express.errorHandler());
});

// Return a 400 response if the combo handler generates a BadRequest error.
app.use(combo.errorHandler());

// Given a root path that points to a YUI 3 root folder, this route will
// handle URLs like:
//
// http://example.com/yui3?build/yui/yui-min.js&build/loader/loader-min.js
//
app.get('/yui3', combo.combine({rootPath: '/local/path/to/yui3'}), combo.respond);

app.listen(3000);
```

#### `combo.respond`

The `respond` method exported by `require('combohandler')` is a convenience
method intended to be the last callback passed to an
[express route](http://expressjs.com/api.html#app.VERB).
Unless you have a *very* good reason to avoid it, you should probably use it.
Here is the equivalent callback:

```js
function respond(req, res) {
    res.send(res.body);
}
```

This method may be extended in the future to do fancy things with optional
combohandler middleware.

#### `combo.errorHandler`

The `errorHandler` export encapsulates the convention of sending `BadRequest`
errors with an optional `errorMaxAge` config.
By default, `BadRequest` errors are served with a 5 minute `max-age` header.

To explicitly disable caching (via
`Pragma: no-cache` and
`Cache-Control: private,no-store`
headers), pass `null` in the options object:

```js
app.use(combo.errorHandler({
    errorMaxAge: null
}));
```

Any other value (including zero) for `errorMaxAge` is interpreted as the
desired duration in seconds.

### Creating a server

If you just want to get a server up and running quickly by specifying a mapping
of routes to local root paths, use the `combohandler/lib/server` module.
It creates a barebones Express server that will perform combo handling on the
routes you specify:

```js
var comboServer = require('combohandler/lib/server'),
    app;

app = comboServer({
    roots: {
        '/yui3': '/local/path/to/yui3'
    }
});

app.listen(3000);
```

### Augmenting an existing server

If you already have an existing Express server instance and just want to add
some combo handled routes to it easily, you can augment your existing server
with combo handled routes:

```js
var comboServer = require('combohandler/lib/server');

comboServer({
    roots: {
        '/yui3': '/local/path/to/yui3'
    }
}, myApp); // Assuming `myApp` is a pre-existing Express server instance.
```

### From the command line

If installed globally via `npm -g install`,
the CLI executable `combohandler` is provided.
If you're operating from a local clone,
`npm link` in the repository root and you're off to the races.
To start the default single-process server,
it's as simple as

```bash
combohandler
# combohandler now running until you hit Ctrl+C
```

Of course, the default output leaves something to be desired: that is to say,
any output.

#### Root Configuration

At the very least,
you need to provide some route-to-rootPath mappings for your CLI combohandler.

When passed in the `--rootsFile` option,
the JSON file contents should follow this pattern:

```json
{
    "/yui3": "/local/path/to/yui3"
}
```

When passed as individual `--root` parameters,
the equivalent to the JSON above looks like this:

```bash
combohandler --root /yui3:/local/path/to/yui3 [...]
```

To run the standalone server in production mode, set the `NODE_ENV` variable to
`production` before running it:

```bash
    NODE_ENV=production combohandler --root /yui3:/path/to/yui3
```

#### CLI Usage

```text
Usage: combohandler [options]

General Options:
  -h, --help        Output this text
  -v, --version     Prints combohandler's version

Combine Options:
  -p, --port        Port to listen on.                                    [8000]
  -a, --server      Script that exports an Express app [combohandler/lib/server]
  -r, --root        String matching the pattern '{route}:{rootPath}'.
                        You may pass any number of unique --root configs.
  -f, --rootsFile   Path to JSON routes config, *exclusive* of --root.
  -b, --basePath    URL path to prepend when rewriting relative url()s.     ['']
  -w, --webRoot     Filesystem path to base rewritten relative url()s from. ['']
                    Use this instead of --basePath when using route parameters.
                    Overrides behaviour of --basePath.
  -m, --maxAge      'Cache-Control' and 'Expires' value, in seconds.  [31536000]
                    Set this to `0` to expire immediately, `null` to omit these
                    headers entirely.

Cluster Options:
  --cluster         Enable clustering of server across multiple processes.
  -d, --pids        Directory where pidfiles are stored.       [$PREFIX/var/run]
  -n, --workers     Number of worker processes.          [os.cpus.length, max 8]
  -t, --timeout     Timeout (in ms) for process startup/shutdown.         [5000]

  --restart         Restart a running master's worker processes.       (SIGUSR2)
  --shutdown        Shutdown gracefully, allows connections to close.  (SIGTERM)
  --status          Logs status of master and workers.
  --stop            Stop server abruptly, not waiting for connections. (SIGKILL)
```

The `--port` and `--server` options may also be set via npm package config settings:

```bash
npm -g config set combohandler:port 2702
npm -g config set combohandler:server /path/to/server.js
```

Unlike the `--server` option, a path specified in this manner *must* be absolute.

### Clustered!

With the advent of `node` v0.8.x, the core `cluster` module is now usable,
and `combohandler` now regains the capability it once had.
Huzzah! said the villagers.

To run a clustered combohandler from the CLI, just add the `--cluster` flag:

```bash
combohandler --cluster --root /yui3:/path/to/yui3
```

To clusterize combohandler from a module dependency,
`combohandler/lib/cluster` is your friend:

```js
var comboCluster = require('combohandler/lib/cluster');
var app = comboCluster({
    pids: '/path/to/piddir',
    server: './myserver.js',
    roots: {
        '/yui3': '/local/path/to/yui3'
    }
});
app.listen(2702);
```

Optional Middleware
-------------------

### Rewriting URLs in CSS files

Because the combo handler changes the path from which CSS files are loaded,
relative URLs in CSS files need to be updated to be relative to the
combohandled path.
Set the `basePath` or `webRoot` configuration option to have the
combohandler default middleware do this automatically.

```js
// This static route can be used to load images and other assets that shouldn't
// be combined.
//
app.use('/public', express.static(__dirname + '/public'));

// This route will combine requests for files in the public directory, and will
// also automatically rewrite relative paths in CSS files to point to the
// non-combohandled static route defined above.
//
app.get('/combo', combo.combine({
    rootPath: __dirname + '/public',
    basePath: '/public'
}), combo.respond);

// The equivalent config as the previous route, except using webRoot
app.get('/combo', combo.combine({
    rootPath: __dirname + '/public',
    webRoot : __dirname
}), combo.respond);
```

Alternatively, you can use the built-in `cssUrls` middleware as a separate
route callback. `cssUrls` must always be placed after the default `combine`
middleware when used in this fashion.

```js
// This route provides the same behaviour as the previous example, providing
// better separation of concerns and the possibility of inserting custom
// middleware between the built-in steps.
app.get('/combo',
    combo.combine({
        rootPath: __dirname + '/public'
    }),
    combo.cssUrls({
        basePath: '/public'
    }),
    combo.respond);
```

Finally, the `cssUrls` middleware has the ability (disabled by default) to
rewrite `@import` paths in the same manner as `url()` values. As `@import` is
considered an anti-pattern in production code, this functionality is strictly
opt-in and requires passing `true` as the `rewriteImports` property in the
middleware options object.

```js
// Automagically
app.get('/combo', combo.combine({
    rootPath: __dirname + '/public',
    webRoot : __dirname,
    rewriteImports: true
}), combo.respond);

// As explicit middleware
app.get('/combo',
    combo.combine({ rootPath: __dirname + '/public' }),
    combo.cssUrls({ basePath: '/public', rewriteImports: true }),
    combo.respond);
```

#### `basePath` or `webRoot`?

In the simplest case,
`basePath` and `webRoot` reach the same result from different directions.
`basePath` allows you to rewrite a single well-known path under any root,
whereas `webRoot` will handle any number of paths under a well-known root.

In general, if you are using both optional middleware,
you should prefer `webRoot` over `basePath`.

### Dynamic Paths via Route Parameters

To enable resolution of dynamic subtree paths under a given `rootPath`,
simply add a [route parameter](http://expressjs.com/api.html#req.params)
to both the route and the `rootPath` config.

```js
app.get('/combo/yui/:version', combo.combine({
    rootPath: '/local/path/to/yui/:version/build'
}), combo.respond);
```

Given this config,
any [YUI release tarball](http://yuilibrary.com/download/yui3/) you explode
into a versioned subdirectory of `/local/path/to/yui/` would be available
under a much shorter URL than the default config provides:

```text
    http://example.com/combo/yui/3.9.1?yui/yui-min.js&yui-throttle/yui-throttle-min.js
    // vs
    http://example.com/combo/yui?3.9.1/build/yui/yui-min.js&3.9.1/build/yui-throttle/yui-throttle-min.js
```

If the built-in `dynamicPath` middleware is used manually, it _must_ be
inserted *before* the default `combine` middleware.


Using as a YUI 3 combo handler
------------------------------

With a tiny bit of configuration, you can tell YUI to use your custom combo
handler instead of the Yahoo! combo handler. Here's an example:

```html
<script src="http://example.com/combo/yui3?build/yui/yui-min.js"></script>
<script>
YUI({
    comboBase: 'http://example.com/combo/yui3?',
    combine  : true,
    root     : 'build/'
}).use('node', function (Y) {
    // YUI will now automatically load modules from the custom combo handler.
});
</script>
```


License
-------

Copyright (c) 2012 Yahoo! Inc. All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
Combo Handler History
=====================

0.4.0 (2014-08-04)
------------------

* Fixed non-whitespace captures to avoid errors when rewriting minified CSS
  imports. (#80)

* Added support for Express 4.7.3+ (thanks @natecavanaugh).

* Updated dependency versions for mkdirp, nopt, and URIjs.

0.3.8 (2013-10-22)
------------------

* Fixed errant nopt config, acknowledging `--no-resolve-symlinks` when passed.

0.3.7 (2013-10-22)
------------------

* Fixed a few `dynamicPath` middleware edge cases with multiple and duplicate
  route parameters.

* Added --no-resolve-symlinks option that uses fs.statSync to validate rootPath
  instead of fs.realpathSync. Caveat emptor.

* Updated jshint devDependency to 2.3.0.

* Updated should devDependency to 2.0.2.

0.3.6 (2013-10-18)
------------------

* The `dynamicPath` middleware now understands multiple parameters in a route.

* Increased test coverage substantially.

* All dependency versions are now explicit.

* Updated express dependency to 3.4.0.

* Updated should devDependency to 2.0.1.

0.3.5 (2013-10-08)
------------------

* Added `errorHandler` middleware to main export, encapsulating a convenient
  way to cache error responses for a different duration than normal responses.

* Ensured that Unicode BOM in the response is removed, if found.
  [elementstorm]

* Fixed tests in `node` v0.11.x,  adding jshint and sinon to `devDependencies`.

* Updated express dependency to 3.3.x. [Eric Ferraiuolo]

0.3.4 (2013-04-29)
------------------

* Encapsulated option-based instance CLI behaviour with args.invoke().

* Exposed CLI option config on the exported object to allow ad-hoc
  customization for those that `require('combohandler/lib/args')`.

0.3.3 (2013-04-26)
------------------

* Fixed a naive regular expression that was preventing url() rewriting when
  background property shorthand syntax was used.

0.3.2 (2013-04-26)
------------------

* Added --webRoot option to support CSS url() rewriting across multiple
  dynamic paths. This option will supplant --basePath if both are present.

* Changed dynamicPath resolution to use fs.stat() instead of fs.realpath(),
  allowing symlinks under a well-known rootPath to build an "absolutely
  relative" pathname instead of always resolving to the rootPath.

* Reorganized tests to cut down on repetitive string concatenation and
  clarify intention. Also added a bunch of complex symlink tests.

0.3.1 (2013-04-24)
------------------

* Fixed cluster worker configuration via events rather than `setupMaster`
  args. Custom `server` config now works, and receives all properties from
  the parsed CLI arguments.

* Changed all pidfile IO to be synchronous. It's not frequent enough to justify
  the indeterminate nature of asynchronicity (which wasn't even used, anyway).

* Added more tests, increasing code coverage slightly.

0.3.0 (2013-04-22)
------------------

* Removed `app.js` and `config.sample.js`. They served us well.

* Changed Travis script to also run code coverage, which is also now checked
  against reasonable threshholds.

* Changed default `lib/server` to consume all available config for each route
  in addition to the `rootPath`.

* Added CLI executable `combohandler` to replace the old `app.js`, among many,
  many other things.

* Added new multi-process implementation built on top of `node` v0.8.x+ core
  `cluster` module.

* Added oodles and oodles of tests.

* Added `mkdirp`, `nopt`, and `rimraf` dependencies.

* Updated `request` and `should` dependency versions.

* Updated package.json config with defaults for `port` and `server` options.

0.2.2 (2013-04-21)
------------------

* Changed default `combine` middleware to return an array of middleware instead
  of a function, allowing addition of bundled middlware depending on configured
  values. (Express flattens any arrays passed as route callbacks)

* Changed default error handler to pass the error to `next()` when it isn't a
  `BadRequest`, which was itself extracted to a separate file.

* Changed CSS `url()` rewriter to use built-in `path` methods instead of custom
  algorithm, with expanded test coverage.

* Added `bodyContents` and `relativePaths` arrays to `res.locals` object in the
  default `combine` middleware, allowing subsequent middleware access to those
  values.

* Added `cssUrls` middleware, an extraction of the CSS `url()` rewriter, with
  the optional capability of rewriting `@import` statements as well.

* Added `dynamicPath` middleware, supporting route params (e.g., `:version`)
  that point to different trees under the same filesystem root.

* Added `respond` middleware for convenience. It simply responds 200 with the
  contents of the `res.body` property.

* Added code coverage with `istanbul`.

* Added Travis support.

* Added [Daniel Stockman](https://github.com/evocateur) as a maintainer.

* Updated mocha dependency to 1.9.0.

* Updated express dependency to 3.2.x. [Eric Ferraiuolo]

0.2.1 (2013-04-01)
------------------

* Added a `basePath` config option that can be used to specify a
  non-combohandled base path. Relative URLs in combohandled CSS files will
  be automatically rewritten to be relative to this base path rather than the
  combo URL. [Ryan Cannon]


0.2.0 (2012-11-03)
------------------

* Supports Express 3.0.

* Removed support for Cluster, since it's dead.

* Added support for symlinks that point to files outside the root path.

* Added support for setting `Cache-Control` and `Expires` headers via the
  `maxAge` config property. [Daniel Stockman]

* Errors now provide more descriptive messages. [Daniel Stockman]

* Deny requests with non-whitelisted or differing MIME types. [Daniel Stockman]

* Return a quick error on requests that are obviously truncated or otherwise
  mangled. [Daniel Stockman]


0.1.3 (2011-10-31)
------------------

* Use Cluster instead of Spark2.


0.1.2 (2011-07-11)
------------------

* Chasing the latest Express and Connect versions again.


0.1.1 (2011-04-12)

lib/args.js
/**
CLI arguments and config cleaning.
**/
var nopt = require('nopt');
var path = require('path');
var fs = require('fs');

var knownOpts = {
    "help": Boolean,
    "version": Boolean,
    "basePath": path,
    "cluster": Boolean,
    "maxAge": Number,
    "pids": path,
    "port": Number,
    "restart": Boolean,
    "resolveSymlinks": Boolean,
    "root": [String, Array],
    "rootsFile": path,
    "server": path,
    "shutdown": Boolean,
    "status": Boolean,
    "stop": Boolean,
    "timeout": Number,
    "webRoot": path,
    "workers": Number
};

var shortHands = {
    "no-resolve-symlinks": ["--no-resolveSymlinks"],
    "h": ["--help"],
    "v": ["--version"],
    "a": ["--server"],
    "b": ["--basePath"],
    "d": ["--pids"],
    "m": ["--maxAge"],
    "p": ["--port"],
    "r": ["--root"],
    "f": ["--rootsFile"],
    "t": ["--timeout"],
    "w": ["--webRoot"],
    "n": ["--workers"]
};

exports = module.exports = {
    knownOpts: knownOpts,
    shortHands: shortHands,
    get usage() {
        var msg = [];

        msg.push("Usage: combohandler [options]");
        msg.push("");
        msg.push("General Options:");
        msg.push("  -h, --help        Output this text");
        msg.push("  -v, --version     Prints combohandler's version");
        msg.push("");
        msg.push("Combine Options:");
        msg.push("  -p, --port      * Port to listen on.                                    [8000]");
        msg.push("  -a, --server    * Script that exports an Express app [combohandler/lib/server]");
        msg.push("  -r, --root        String matching the pattern '{route}:{rootPath}'.");
        msg.push("                        You may pass any number of unique --root configs.");
        msg.push("  -f, --rootsFile   Path to JSON routes config, *exclusive* of --root.");
        msg.push("  -b, --basePath    URL path to prepend when rewriting relative url()s.     ['']");
        msg.push("  -w, --webRoot     Filesystem path to base rewritten relative url()s from. ['']");
        msg.push("                    Use this instead of --basePath when using route parameters.");
        msg.push("                    Overrides behaviour of --basePath.");
        msg.push("  -m, --maxAge      'Cache-Control' and 'Expires' value, in seconds.  [31536000]");
        msg.push("                    Set this to `0` to expire immediately, `null` to omit these");
        msg.push("                    headers entirely.");
        msg.push("");
        msg.push("  --no-resolve-symlinks");
        msg.push("      If passed, rootPaths that are symlinks will not be resolved (the default).");
        msg.push("");
        msg.push("Cluster Options:");
        msg.push("  --cluster         Enable clustering of server across multiple processes.");
        msg.push("  -d, --pids        Directory where pidfiles are stored.       [$PREFIX/var/run]");
        msg.push("  -n, --workers     Number of worker processes.          [os.cpus.length, max 8]");
        msg.push("  -t, --timeout     Timeout (in ms) for process startup/shutdown.         [5000]");
        msg.push("");
        msg.push("  --restart         Restart a running master's worker processes.       (SIGUSR2)");
        msg.push("  --shutdown        Shutdown gracefully, allows connections to close.  (SIGTERM)");
        msg.push("  --status          Logs status of master and workers.");
        msg.push("  --stop            Stop server abruptly, not waiting for connections. (SIGKILL)");
        msg.push("");
        msg.push("* The port and server options can also be set via npm package config settings:");
        msg.push("    npm -g config set combohandler:port 2702");
        msg.push("    npm -g config set combohandler:server /path/to/server.js");
        msg.push("");
        msg.push("combohandler@" + exports.version);

        return msg.join('\n');
    },
    get version() {
        return require('../package.json').version;
    },
    clean: function (config) {
        nopt.clean(config, knownOpts);

        return resolveRoots(config);
    },
    parse: function (args, slice) {
        var config = nopt(knownOpts, shortHands, args, slice);

        // allow one string argument to stand in for Boolean flag
        config.argv.remain.some(function (arg) {
            /*jshint boss: true */
            switch (arg) {
            case 'restart':
                return config.restart = true;
            case 'shutdown':
                return config.shutdown = true;
            case 'status':
                return config.status = true;
            case 'stop':
                return config.stop = true;
            }
        });

        return resolveRoots(config);
    },
    /**
    Encapsulates behaviour of an instance when created via CLI.

    @method invoke
    @param {ComboCluster} instance
    **/
    invoke: function (instance) {
        var options = instance.options;
        if (options.restart) {
            instance.restart();
        }
        else if (options.shutdown) {
            instance.shutdown();
        }
        else if (options.status) {
            instance.status();
        }
        else if (options.stop) {
            instance.stop();
        }
        else {
            instance.listen();
        }
    }
};

function resolveRoots(config) {
    // parse roots json config
    if (config.rootsFile) {
        var rootStat = fs.statSync(config.rootsFile);
        if (rootStat.isFile()) {
            config.roots = require(config.rootsFile);
            var rootDir = path.dirname(config.rootsFile);

            // resolve rootDir to route paths
            Object.keys(config.roots).forEach(function (route) {
                config.roots[route] = path.resolve(rootDir, config.roots[route]);
            });
        }
    } else if (config.root) {
        // individual mapping passed as
        // --root /route:/local/path
        config.roots = {};

        config.root.forEach(function (route) {
            route = route.split(':');

            // route might have parameters in it (:foo)
            if (route.length > 2) {
                // parameters are always matched, so length is always even
                var idx = route.length,
                    step = idx / 2;

                // loop in reverse
                while (idx) {
                    // only twice (second decrement makes it zero)
                    idx -= step;
                    // joining each half again with colons
                    route.splice(idx, 0, route.splice(idx, step).join(':'));
                }
            }

            config.roots[route[0]] = path.resolve(route[1]);
        });
    }

    return config;
}
